job-server-api/src/spark.jobserver/NamedObjectSupport.scala

```diff
@@ -0,0 +1,134 @@
+package spark.jobserver
+
+import akka.util.Timeout
+import java.lang.ThreadLocal
+import java.util.concurrent.atomic.AtomicReference
+import org.apache.spark.rdd.RDD
+import org.apache.spark.SparkContext
+
+import org.apache.spark.sql.DataFrame
+import org.apache.spark.storage.StorageLevel
+import scala.concurrent.duration.Duration
+
+sealed trait NamedObject
+case class NamedRDD[T](rdd: RDD[T], storageLevel: StorageLevel) extends NamedObject
+case class NamedDataFrame(df: DataFrame) extends NamedObject
+
+abstract class NamedObjectPersister[O <: NamedObject] {
+  //TODO what is the context needed for? c: ContextLike,
+  def saveToContext(namedObj: O, name: String): Unit
+  //TODO - what is this for?
+  //def getFromContext(c: ContextLike, name: String): O
+  /**
+   * update reference to named object so that it does not get GCed
+   */
+  def refresh(namedObject: O): O
+  //TODO - what about 'unpersist' ?
+}
+
+/**
+ * NamedObjects - a trait that gives you safe, concurrent creation and access to named objects
+ * such as RDDs or DataFrames
+ * (the native SparkContext interface only has access to RDDs by numbers).
+ * It facilitates easy sharing of data objects amongst jobs sharing the same SparkContext.
+ * If two jobs simultaneously tries to create a data object with the same name and in the same namespace,
+ * only one will win and the other will retrieve the same one.
+ *
+ * Note that to take advantage of NamedObjectSupport, a job must mix this in and use the APIs here instead of
+ * the native DataFrame/RDD `cache()`, otherwise we will not know about the names.
+ */
+trait NamedObjects {
+  // Default timeout is 60 seconds. Hopefully that is enough
+  // to let most RDD/DataFrame generator functions finish.
+  val defaultTimeout = akka.util.Timeout(Duration(60, java.util.concurrent.TimeUnit.SECONDS))
```

Comment:
ok, introduced `spark.jobserver.named-object-creation-timeout`
However, to get this to work I had to change the timeout type to `Option[Timeout]` as in: 
    def getOrElseCreate[O <: NamedObject](name: String,
                         objGen: => O)(implicit timeout: Option[Timeout] = None,
                                       persister: NamedObjectPersister[O]): O